File: safety_node2/src/safety_node.py
#!/usr/bin/env python
import rospy

# TODO: import ROS msg types and libraries

import numpy as np
import precompute
from std_msgs.msg import String,Bool
from Ackermann_msgs.msg import AckermannDriveStamped
import fstream
from nav_msgs.msg import Odometry
from sensor_msgs.msg import LaserScan


class Safety(object):
    """
    The class that handles emergency braking.
    """
    def __init__(self):
        """
        One publisher should publish to the /brake topic with a AckermannDriveStamped brake message.

        One publisher should publish to the /brake_bool topic with a Bool message.

        You should also subscribe to the /scan topic to get the LaserScan messages and
        the /odom topic to get the current speed of the vehicle.

        The subscribers should use the provided odom_callback and scan_callback as callback methods

        NOTE that the x component of the linear velocity in odom is the speed
        """
        self.speed = 0
        # TODO: create ROS subscribers and publishers.
        rospy.Subscriber("odom",Odometry,odom_callback)
        rospy.Subscriber("scan",LaserScan,scan_callback)
        
        brake_bool_pub=rospy.Publisher("brake_bool",Bool,queue_size=1)
        brake_pub=rospy.Publisher("brake",AckermannDriveStamped,queue_size=1)
        
        self.brake_ttc_threshold=rospy.get_param('~brake_ttc_threshold')
        
        scan_beams=rospy.get_param('~scan_beams')
        wheelbase=rospy.get_param('~wheelbase')
        width=rospy.get_param('~width')
        scan_dist_to_base_link=rospy.get_param('~scan_distance_to_base_link')
        scan_fov=rospy.get_param('~scan_field_of_view')
        scan_angle_incr=scan_fov/scan_beams
        
        precompute=precompute(scan_beams, 
		              wheelbase, 
		              width, 
		              scan_distance_to_base_link,
		              -scan_fov/2.0,
		              scan_angle_incr)
		              
        self.cosines,self.car_distances=precompute.cosines_and_car_dist()
        

    def odom_callback(self, odom_msg):
        # TODO: update current speed
        self.speed = odom_msg.twist.twist.linear.x

    def scan_callback(self, scan_msg):
        # TODO: calculate TTC
        
        brk=False
        
        if(self.speed!=0):
        	for i in range(scan_msg.ranges.size()):
        		proj_speed=self.speed * self.cosines[i]
        		ttc=(scan_msg.ranges[i]-self.car_distances[i])/proj_speed
        		
        		# TODO: publish brake message and publish controller bool
        		
        		if((ttc<self.brake_ttc_threshold) and (ttc>0)):
        			ack_msg=AckermannDriveStamped()
        			ack_msg.header.stamp=rospy.Time.now()
        			ack_msg.drive.steering_angle=0.0
        			ack_msg.drive.speed=0.0
        			brake_pub.publish(ack_msg)
        			
        			brk=True
        			brake_bool.Publish(brk)
        			
        			break
	 
	  #publish brk=False otherwise
	  brake_bool.Publish(brk)
        			
          return
  
def main():
    rospy.init_node('safety_node')
    sn = Safety()
    rospy.spin()
if __name__ == '__main__':
    main()

File: safety_node1/src/safety_node.cpp
// #include <ros/ros.h>
// #include <iostream>
// #include "std_msgs/String.h"
// #include <sstream>
// #include <nav_msgs/Odometry.h>
// #include <sensor_msgs/LaserScan.h>
// #include <ackermann_msgs/AckermannDriveStamped.h>
// // TODO: include ROS msg type headers and libraries

// class Safety {
// // The class that handles emergency braking
// private:
//     ros::NodeHandle n;
//     double speed;
//     // TODO: create ROS subscribers and publishers

//     // Odometry Subscriber
//     ros::Subscriber odom_sub;
//     // LiDar Subscriber
//     ros::Subscriber laser_sub;

//     // Break Publisher
//     ros::Publisher break_bool_pub;
//     ros::Publisher break_pub;

//     // Calculation to compensate the position of the laser scan
//     // Get params for precomputation and collision detection
//     int scan_beams;                     // number of beams 1080
//     double scan_field_of_view;          //
//     double scan_ang_incr;
//     double wheelbase;
//     double width;
//     double scan_distance_to_base_link;
//     double brake_ttc_threshold;         // for collision detection


// public:
//     Safety() {

//         // Initilization Params
//         n = ros::NodeHandle();
//         speed = 0.0;
        
//         n.getParam("brake_ttc_threshold", brake_ttc_threshold);
//         n.getParam("scan_beams", scan_beams);
//         n.getParam("scan_distance_to_base_link", scan_distance_to_base_link);
//         n.getParam("width", width);
//         n.getParam("wheelbase", wheelbase);
//         n.getParam("scan_field_of_view", scan_field_of_view);
//         scan_ang_incr = scan_field_of_view / scan_beams;
//         /*
//         One publisher should publish to the /brake topic with an
//         ackermann_msgs/AckermannDriveStamped brake message.

//         One publisher should publish to the /brake_bool topic with a
//         std_msgs/Bool message.

//         You should also subscribe to the /scan topic to get the
//         sensor_msgs/LaserScan messages and the /odom topic to get
//         the nav_msgs/Odometry messages

//         The subscribers should use the provided odom_callback and 
//         scan_callback as callback methods

//         NOTE that the x component of the linear velocity in odom is the speed
//         */

//         // TODO: create ROS subscribers and publishers

//         // Start subscribers to listen to laser scan and odom messages
//         laser_sub = n.subscribe("scan",1,&Safety::scan_callback,this);
//         odom_sub = n.subscribe("odom",1,&Safety::odom_callback,this);

//         // Make a publisher for brake messages
//         break_bool_pub = n.advertise<std_msgs>("brake_bool",1);
//         break_pub = n.advertise<ackerm
//     }
//     void odom_callback(const nav_msgs::Odometry::ConstPtr &odom_msg) {
//         // TODO: update current speed
//         double x_velocity = odom_msg->twist.twist.linear.x;
//         speed = x_velocity;
//     }

//     void scan_callback(const sensor_msgs::LaserScan::ConstPtr &scan_msg) {
        
//         // get Laser-scan Infos for 
//         std::vector<float> ->ranges
//         // TODO: calculate TTC


//         // TODO: publish drive/brake message
//     }

// };

// int main(int argc, char ** argv){
//      //Initialize and start the node
//      ros::init(argc, argv, "safety_node");
//      ros::NodeHandle n;
//      ROS_INFO_STREAM("starting with safety brake") ;
//     std::cout << "safety_node node initialised" << std::endl;
//     //handle ROS communication events
//     /* The second parameter to advertise() is the size of the message queue
//       * used for publishing messages.  If messages are published more quickly
//       * than we can send them, the number here specifies how many messages to
//       * buffer up before throwing some away.
//         */
//     ros::Publisher chatter_pub = n.advertise<std_msgs::String>("chatter", 1000);  
//     ros::Rate loop_rate(10);
//     int count=0;

//     while (ros::ok())
//     {
//         std_msgs::String msg;

//         std::stringstream ss;
//         ss << "hello" << count;
//         msg.data=ss.str();

//         ROS_INFO("%s", msg.data.c_str());

//         chatter_pub.publish(msg);
//         ros::spinOnce();
//         loop_rate.sleep();
//         count++;

//     }
    
//     return 0;
// }
